src/ev3_bluetooth.c

```diff
@@ -0,0 +1,332 @@
+#include "../firmware_headers/bluetooth/hci_lib.h"
+#include "../firmware_headers/bluetooth/rfcomm.h"
+
+#include "ev3_bluetooth.h"
+
+/**
+ * Channel 1 is already taken by the UI process of the robot, so we use 2.
+ */
+#define DEFAULT_C4EV3_RFCOMM_CHANNEL    2
+#define MAX_CONNECTIONS                 7
+#define SHORT_BLUETOOTH_ADDRESS_LENGTH  sizeof("000000000000")
+
+#define KNOWN_BLUETOOTH_NAMES_FILE_NAME_LENGTH 256
+#define KNOWN_BLUETOOTH_NAMES_FILE_ROW_LENGTH  256
+
+void BluetoothInit () { }
+
+BluetoothConnectionHandle connections[MAX_CONNECTIONS];
+int currentConnections = 0;
+
+void addConnectionToList (BluetoothConnectionHandle c) {
+	if (currentConnections > MAX_CONNECTIONS) {
+		return;
+	}
+	connections[currentConnections++] = c;
+}
+
+void removeConnectionFromList (BluetoothConnectionHandle c) {
+	int i;
+	for (i = 0; i < currentConnections; i++) {
+		if (connections[i] == c) {
+			break;
+		}
+	}
+	if (i >= currentConnections) {
+		return;
+	}
+	currentConnections--;
+	for (; i < currentConnections; i++) {
+		connections[i] = connections[i + 1];
+	}
+}
+
+
+bool isBluetoothAddress(const char * nameOrAddress);
+bool findAddressByBluetoothName(const char * name, char * address);
+
+/**
+ * Connects to the specified bluetooth address. If the connection fails, a new
+ * attempt is tried after one second.
+ * @param address
+ * @param attempts number of attempts to retry connection
+ * @return connection handle or -1 if the connection couldn't be established
+ */
+BluetoothConnectionHandle connectByBluetoothAddress(const char * address, int attempts);
+
+BluetoothConnectionHandle ConnectTo(const char * nameOrAddress, int attempts) {
+	char address[BLUETOOTH_ADDRESS_LENGTH];
+	if (isBluetoothAddress(nameOrAddress)) {
+		strncpy(address, nameOrAddress, BLUETOOTH_ADDRESS_LENGTH);
+		address[BLUETOOTH_ADDRESS_LENGTH - 1] = '/0';
+	} else {
+		bool found = findAddressByBluetoothName(nameOrAddress, address);
+		if (!found) {
+			return -1;
+		}
+	}
+	BluetoothConnectionHandle c = connectByBluetoothAddress(address, attempts);
+	addConnectionToList(c);
+	return c;
+}
+
+bool isBluetoothAddress(const char * nameOrAddress) {
+    return bachk(nameOrAddress) == 0;
+}
+
+FILE * openKnownBluetoothNamesFile(void);
+void getLocalBluetoothAddress (char * address);
+void getKnownBluetoothNamesFile (const char * localBluetoothAddress, char * fileName);
+bool getAddressFromKnownBluetoothNamesFileRow(const char *addressAndName, char * address);
+void getNameFromKnownBluetoothNamesFileRow(const char *addressAndName, char * name);
+
+/**
+ * To find the address of a device, given its name, we read a file that the
+ * firmware of the robot writes repeatedly. This file contains a line for each
+ * known bluetooth device. Each line contains the bluetooth address and name.
+ * We don't use the BlueZ C API directly because that part of the API causes the
+ * UI of the robot to crash.
+ * @param nameToFind
+ * @param foundAddress
+ * @return
```

Comment:
I like you adding Function documentation. Would be great if you can fill out the return clause as well. :)

Reply:
7e0be6d